### PR TITLE
Text edit mode + Moveable resizing/rotation

### LINKED ISSUE #1801: Text element: Resizing/rotating should be possible in edit mode
## Feature Description

By original design, the Movable features (such as resizing and rotation) were disabled for a Text element in edit mode. The design has now been updated and it should be possible to scale and change the width of a text element in edit mode as well.

[See Figma](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=8011%3A14748).

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria
1. It should be possible to scale and change the width of a Text element in editor mode
2. It should be possible to rotate an element in edit mode
3. The font size should be adjusted while resizing, just as in the display mode

### PR DESCRIPTION
**Pull request by miina**
## Summary

Adds `SingleSelectionMoveable` to Text edit mode, reuses the same logic mostly with some special handling.

## Relevant Technical Choices
- The Moveable/element wrapper needs to be visually updated when the text is changing the height of the element while being in edit mode.
- If the element height is updated due to doing things in edit mode, and then the user wants to resize the element, the element properties need to be updated before the user can do that -- otherwise, the calculations will be based on the original element.
- Dragging is not allowed in edit mode, however, we can't just disable it since otherwise selecting part of the text will be disabled.

## To-do

- [x] Tests

## User-facing changes

The user can now resize and rotate the text element while in edit mode.

## Testing Instructions

I
1. Add a Text element and enter edit mode.
2. Verify that resizing and rotation is enabled in edit mode as well and the element is rotated/resized as expected.

II
1. Add a Text element and enter edit mode.
2. Add some empty lines / new lines and some text to change the element height while in edit mode.
3. Verify that the blue line for resizing/rotation is following the new text height.
4. Now, without exiting the edit mode, resize/rotate the element.
5. Verify that the resizing/rotation is working as expected.

---

Fixes #1801

### CODE REVIEW

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 1]
```diff
@@ -288,6 +300,9 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
         if (isResizingFromCorner !== newResizingMode) {
           setIsResizingFromCorner(newResizingMode);
         }
+        if (isEditMode) {
+          setActionHappening(true);
```
Comment (miina) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js]: This is needed here since otherwise, if we change the height of the element in the edit mode via changing the text (e.g. by adding some lines), the element will be out of sync and Moveable will not work as expected. So we need to signal the edit mode to update the element properties for these being in sync.

Not sure if there's a better alternative for this.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 2]
```diff
@@ -252,9 +267,15 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
       }}
       throttleDrag={0}
       onDragStart={({ set }) => {
+        if (isEditMode) {
+          return false;
```
Comment (miina) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js:272]: Why not just disable dragging?
When disabling dragging, Moveable dismisses any events and it's not possible to click or select text inside the edit box, even if the text edit is on top of moveable. Enabling dragging but returning `false` allows events but not dragging.

File: assets/src/edit-story/elements/text/edit.js [hunk 3]
```diff
@@ -205,13 +196,29 @@ function TextEdit({
   // Update content for element on unmount.
   useUnmount(updateContent);
 
+  useEffect(() => {
+    // If there are any moveable actions happening, let's update the content
+    // Otherwise the font size and measures will not be correct.
+    if (actionHappening) {
+      updateContent();
```
Comment (miina) [inline assets/src/edit-story/elements/text/edit.js:208]: Here we update the content so that Moveable could do its calculations based on the updated element.

File: assets/src/edit-story/elements/text/edit.js [hunk 4]
```diff
@@ -205,13 +196,29 @@ function TextEdit({
...
     const wrapper = wrapperRef.current;
     const textBox = textBoxRef.current;
     editorHeightRef.current = textBox.offsetHeight;
     wrapper.style.height = `${editorHeightRef.current}px`;
-  }, []);
+    if (editWrapper) {
+      // @todo Get the correct top and left from getBox, just height for testing currently!
+      editWrapper.style.height = `${editorHeightRef.current}px`;
+      // We need to update Moveable as well.
+      if (moveable.current) {
```
Comment (miina) [inline assets/src/edit-story/elements/text/edit.js]: The height has been updated (the element properties not yet), we need to visually update Moveable as well to reflect that.

File: assets/src/edit-story/components/canvas/editLayer.js [hunk 5]
```diff
@@ -81,6 +83,9 @@ function EditLayerForElement({ element }) {
     return () => focusCanvas(/* force */ false);
   }, [focusCanvas]);
 
+  // Display the overflow in case the element supports grayout or movable:
+  // Both need overflow to work as expected.
+  const showOverFlow = editModeGrayout || hasEditModeMovable;
```
Comment (miina) [inline assets/src/edit-story/components/canvas/editLayer.js]: From Dima's comment: this can always show overflow, will update.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 6]
```diff
@@ -40,8 +40,18 @@ const VERTICAL_HANDLES = ['n', 's'];
...
+  selectedElement,
+  targetEl,
+  pushEvent,
+  isEditMode,
+  editMoveableRef,
+  setActionHappening,
+}) {
+  let moveable = useRef(null);
+  if (isEditMode) {
+    moveable = editMoveableRef;
```
Comment (miina) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js]: Here we're using the ref that was passed from the edit mode -- so that we could update the Moveable if the height has changed in the edit mode without the element properties being updated yet.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 7]
```diff
@@ -235,6 +247,9 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
       resizable={actionsEnabled && !hideHandles}
       rotatable={actionsEnabled && !hideHandles}
       onDrag={({ target, beforeTranslate, clientX, clientY }) => {
+        if (isEditMode) {
+          return false;
```
Comment (miina) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js:251]: (See the next comment about why not just disable dragging)

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 1 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: Perhaps an alternative could be tracking the original height within the Text Edit mode and updating the properties of the Text element every time when there's a difference with the previous height? This way it would always be up to date for Moveable.

This would cause a new history entry for each edit which causes a height change though, also not the best option.

File: assets/src/edit-story/components/canvas/editElement.js [hunk 8]
```diff
@@ -44,17 +47,47 @@ function EditElement({ element }) {
     getBox: state.actions.getBox,
   }));
 
-  const { Edit } = getDefinitionForType(type);
+  const [editWrapper, setEditWrapper] = useState(null);
+  const [actionHappening, setActionHappening] = useState(false);
+
+  const { lastSelectionEvent } = useCanvas(
```
Comment (dvoytenko) [inline assets/src/edit-story/components/canvas/editElement.js]: `lastSelectionEvent`, I believe, is used only explicitly for dragging. Since we're disabling it, we should drop it from here as well.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 6 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: I think we had combinedRefs code somewhere?

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 1 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: I believe the movable only tracks the outline of the target DOM element. So all we need to do to keep them in sync is to call `moveable.updateRect()`. It's already called in the `SingleSelectionMovable`, but since it's outside of the rendering path it's probably doesn't get called. So we can either call `updateRect()` manually, or we could find a way to get the whole `SingleSelectionMovable` to rerender.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 9]
```diff
@@ -329,6 +353,9 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
         setTransformStyle(target);
       }}
       onResizeEnd={({ target }) => {
+        if (isEditMode) {
+          setActionHappening(false);
```
Comment (dvoytenko) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js]: We already should have `setActionHappening` via either `useTransformHandler` or `isAnythingTransforming`.

File: assets/src/edit-story/elements/text/edit.js [hunk 10]
```diff
@@ -45,28 +45,14 @@ import useUnmount from '../../utils/useUnmount';
...
 import { generateParagraphTextStyle, getHighlightLineheight } from './util';
 
 // Wrapper bounds the text editor within the element bounds. The resize
 // logic updates the height of this element to show the new height based
 // on the content and properties.
-// Background color is used to make the edited element more prominent and
-// easier to see.
 const Wrapper = styled.div`
   ${elementFillContent}
-  background-color: ${({ theme }) => theme.colors.whiteout};
```
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/edit.js:57]: This, imho, belongs in a separate pull request.

File: assets/src/edit-story/elements/text/edit.js [hunk 11]
```diff
@@ -232,6 +239,14 @@ function TextEdit({
     ]);
   }, [font, fontFaceSetConfigs, maybeEnqueueFontStyle]);
 
+  useTransformHandler(id, (transform) => {
+    const target = textBoxRef.current;
+    const updatedFontSize = transform?.updates?.fontSize;
```
Comment (dvoytenko) [inline assets/src/edit-story/elements/text/edit.js:255]: It might be better for this component to compute the `fontSize` itself, and possibly to turn off that feature in the `SingleSelectionMovable`.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 1 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: It was actually the `updateForResizeEvent` used for the text element which didn't work as expected unless the element was updated.

Additionally, it might make sense to update the element before each resizing for adding the text editing that happened before resizing as a separate history entry. Thoughts?

Without updating the element properties before doing resizing, the following happened:
1. In edit mode, I added/removed a few extra lines
2. Since `updateRect()` is happening in the text edit mode, the moveable follows the height change correctly.
3. Now when I resize the element, the inner content and the Moveable go out of sync.

Example:
resizing-edit

This doesn't happen if the element properties are updated before resizing (as currently via the `actionsHappening`). Thoughts?

File: assets/src/edit-story/elements/text/edit.js (hunk 10 above)
Comment (miina) [reply assets/src/edit-story/elements/text/edit.js:57]: Leaving this in would require ensuring the Wrapper is also moving together with the Moveable (currently it's not) and would need extra handling for ensuring that.

Since we needed to remove this anyway, it seemed to make sense to remove this here as opposed to adding additional support for it here to just remove this again in #2040 

Thoughts?

File: assets/src/edit-story/elements/text/edit.js (hunk 11 above)
Comment (miina) [reply assets/src/edit-story/elements/text/edit.js:255]: Curious to know the reasoning.
Currently, it's exactly matching the font size in the display -- tested out two equal elements when resizing in display mode vs edit mode.

Also, if the font calculation would be done separately here, wouldn't the history entries include mismatching data? E.g. the height/width/left/top would be set via Moveable to a history entry but the font-size would be added as it was before resizing since it wouldn't be calculated and saved together with the size and position. This would include a history entry with font size out of sync compared to the other props. Thoughts?

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 9 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: `isAythingTransforming` does not work as expected in this case since we actually need to update the element _before_ anything has started transforming. Will check if there's a way how to modify the logic to work for this case, too.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 1 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js]: Using `isAnythingTransforming` now.

File: assets/src/edit-story/components/canvas/pagemenu/index.js [hunk 12]
```diff
@@ -186,7 +186,7 @@ function PageMenu() {
           <WithTooltip title={__('Duplicate page', 'web-stories')}>
             <Icon
               onClick={handleDuplicatePage}
-              aria-label={__('Dupliccate Page', 'web-stories')}
+              aria-label={__('Duplicate Page', 'web-stories')}
```
Comment (swissspidy) [inline assets/src/edit-story/components/canvas/pagemenu/index.js]: Ouch! Good catch. Since it's unrelated, perhaps move it into a separate PR?

Comment (miina) [reply assets/src/edit-story/components/canvas/pagemenu/index.js]: Added: https://github.com/google/web-stories-wp/pull/2905

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js [hunk 13]
```diff
@@ -196,7 +203,9 @@ function SingleSelectionMovable({ selectedElement, targetEl, pushEvent }) {
   );
 
   const canSnap =
-    !snapDisabled && (!isDragging || (isDragging && !activeDropTargetId));
+    !snapDisabled &&
+    !isEditMode &&
+    (!isDragging || (isDragging && !activeDropTargetId));
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/singleSelectionMovable.js:211]: Yeah, this one is definitely getting complex. It would make sense with a cleaner API for this.

File: assets/src/edit-story/elements/text/edit.js (hunk 4 above)
Comment (barklund) [reply assets/src/edit-story/elements/text/edit.js]: Maybe it would be better as a callback, e.g. `onResize`? And then handle the updating in `EditElement`?

**Review by barklund — COMMENTED** [review]
Overall this looks really good. There's a bit of jumping when moving in and out of edit-mode (when in a scaled canvas, might not occur in full-size canvas):

editmode-jump

Also, one major nit: The spellings _Movable_ and _Mov**e**able_ is used completely at random (but that was probably also the case before this PR). We really should settle on a single spelling!

File: assets/src/edit-story/elements/text/edit.js (hunk 4 above)
Comment (miina) [reply assets/src/edit-story/elements/text/edit.js]: Hmm, here the updating of the element is specifically related to Text edit mode only -- changing text is causing the resize. It seems to make more sense to have it here since it seems to be very specific to Text edit, even if theoretically in the future another element would have moveable + edit mode together. Thoughts?

**Comment by miina** [comment]
>Also, one major nit: The spellings Movable and Moveable is used completely at random (but that was probably also the case before this PR). We really should settle on a single spelling!

Good point, changed everything to Moveable since that's the original name of the library we're using, also its classnames are using `moveable` meaning that if we'd switch to `movable` we'd still occasionally need to use `moveable` for styling -- with `moveable` we can fully unify the naming.

File: .github/CODEOWNERS [hunk 14]
```diff
@@ -64,9 +64,9 @@ CONTRIBUTING.md                                                   @swissspidy
 #assets/src/edit-story/elements/image/                             @cvolzke4 @diegovar @jhtin @joannag6 @winterDroid
 #assets/src/edit-story/elements/video/                             @cvolzke4 @diegovar @jhtin @joannag6 @winterDroid
 
-## Movable
-#assets/src/edit-story/components/movable/                         @miina
-#assets/src/edit-story/elements/image/editCropMovable.js           @miina
-#assets/src/edit-story/elements/shared/editPanMovable.js           @miina
-#assets/src/edit-story/components/canvas/singleSelectionMovable.js @miina
-#assets/src/edit-story/components/canvas/multiSelectionMovable.js  @miina
+## Moveable
```
Comment (miina) [inline .github/CODEOWNERS:67]: @swissspidy -- FI: changed these based on renaming all `movable` to `moveable` for consistency. These paths are actually a bit outdated anyway (e.g. there's only the image and not video/media folder), however, didn't change this since not really related to this PR.

File: assets/src/edit-story/app/index.js [hunk 15]
```diff
@@ -47,6 +47,9 @@ import { useSnackbar, SnackbarProvider } from './snackbar';
 import Layout from './layout';
 import { Media3pApiProvider } from './media/media3p/api';
 
+/**
+ *
+ */
```
Comment (miina) [inline assets/src/edit-story/app/index.js:52]: Hmm, looks like all the JSDoc comments were added automatically by the local linter. 🤔

File: assets/src/edit-story/elements/shared/index.js [hunk 16]
```diff
@@ -42,7 +42,7 @@ export const elementWithPosition = css`
 
 export const elementWithSize = css`
   width: ${({ width }) => `${width}px`};
-  height: ${({ height }) => `${height}px`};
+  height: ${({ height }) => `${Math.round(height)}px`};
```
Comment (miina) [inline assets/src/edit-story/elements/shared/index.js]: @barklund This seems to fix the jumpiness caused by inconsistencies between edit-display mode, looks like it happens when height is adjusted due to some related value changing, e.g. font size / line-height which can cause a difference due to using partial vs full pixels. Thoughts?

Comment (miina) [reply assets/src/edit-story/elements/shared/index.js]: Note: this seems to also be necessary for https://github.com/google/web-stories-wp/pull/3125

File: assets/src/edit-story/app/index.js (hunk 15 above)
Comment (swissspidy) [reply assets/src/edit-story/app/index.js:52]: Most warnings should go away with a rebase

File: assets/src/edit-story/elements/text/edit.js (hunk 4 above)
Comment (barklund) [reply assets/src/edit-story/elements/text/edit.js]: It's mostly the fact, that the `TextEdit` component now for weird reasons (as a component) expects an argument, which is a ref to something with an `updateRect` method. That's a rather weirdly specific requirement for this component, that doesn't really make sense if this component is used anywhere else (not that I can think of where). For the sake of SoC, it makes more sense to simply pass a `handleResize` and invoke it inside any editable component (images and videos might be able to resize as well).

**Review by barklund — APPROVED** [review]
Looks good to me, pending the few remaining comment threads.

File: assets/src/edit-story/elements/text/edit.js (hunk 4 above)
Comment (miina) [reply assets/src/edit-story/elements/text/edit.js]: Makes sense, moved the moveable part to `onResize`, PTAL: 15f9a26

Moving to QA meanwhile.

File: assets/src/edit-story/components/canvas/singleSelectionMovable.js (hunk 13 above)
Comment (miina) [reply assets/src/edit-story/components/canvas/singleSelectionMovable.js:211]: Should improve in #3016

**Review by dvoytenko — APPROVED** [review]
